### PR TITLE
Make it correctly

### PR DESCRIPTION
**Pull request by mmlb**
## Description

Fixes to Makefile/build system so parallel builds work better.

## Why is this needed

Parallel builds were previously racy and/or wasting time and bandwidth. We could have either been fetching ipxe source files multiple times (potentially once for each target) or the first build_ipxe.sh to run will start to fetch the tarball and subsequent ones could see the incomplete file and begin to try and extract it with all sorts of fun ensuing.

## How Has This Been Tested?

I've forced fetching/rebuilding of the ipxe files multiple times with no problems. I've also interrupted both curl and tar and verified that next invocations do the right thing.

## How are existing users impacted? What migration steps/scripts do we need?

More reliable builds.

## Checklist:

I have:

- [ ] updated the documentation and/or roadmap (if required)
- [ ] added unit or e2e tests
- [ ] provided instructions on how to upgrade

### CODE REVIEW

**Review by nshalman — CHANGES_REQUESTED** [review]
We already have `binary/script/shell.nix`

Should we relocate it? symlink it?
I don't think we should have two shell.nix files in a single repo if it can be avoided unless there's a very good reason...

**Comment by mmlb** [comment]
> We already have `binary/script/shell.nix`
> 
> Should we relocate it? symlink it? I don't think we should have two shell.nix files in a single repo if it can be avoided unless there's a very good reason...

Yeah I usually agree with that. I tried hoisting to top-level and also updating the nixpkgs but then ipxe failed to build (which brought back lots of memories I had suppressed :smile: ) due to gcc 9 -> 10. Let me try again with updated-nixpkgs+NoCC+gcc9.

**Comment by mmlb** [comment]
I also noticed that this PR is still broken for parallel builds because each target build mutates the same source tree... going to need a copy.

**Comment by mmlb** [comment]
@nshalman PTAL I've been able to make it so we only have 1 shell.nix file.

**Comment by jacobweinstock** [comment]
what do you mean by "Parallel builds"? Where/when are "Parallel builds" being run?

**Comment by mmlb** [comment]
> what do you mean by "Parallel builds"? Where/when are "Parallel builds" being run?

I have `MAKEFLAGS=-j$(nproc)` in my user env, so make builds the targets in parallel when I wanted to test something out locally in an ipxe build.

**Comment by jacobweinstock** [comment]
> > what do you mean by "Parallel builds"? Where/when are "Parallel builds" being run?
> 
> I have `MAKEFLAGS=-j$(nproc)` in my user env, so make builds the targets in parallel when I wanted to test something out locally in an ipxe build.
> 
> More info (including rough timing numbers) can be seen in the commit message for [ce49e25](https://github.com/tinkerbell/ipxedust/commit/ce49e252ac3547b234f36b2987183e58d642d48f)

Ah i see. thanks for the explanation.

**Comment by jacobweinstock** [comment]
Would you be open to leaving the `shell.nix` in its current location (`binary/script/shell.nix`)? It's only needed for iPXE builds. It might be a good idea to also move all the iPXE build related stuff in the current `Makefile` to the same location and then just import that into the top level `Makefile`. I feel like that could break down the complexity of the different moving parts of this repo. Thoughts?

**Comment by mmlb** [comment]
> Would you be open to leaving the `shell.nix` in its current location (`binary/script/shell.nix`)? It's only needed for iPXE builds. 

An earlier version of this PR was like that, but it was still adding a top-level shell.nix file with `curl gnumake gnused go`. If we did the 2 shell.nix files we'd still need gnumake and go, and running the `fetch_and_extract_ipxe.sh` script should be done using nix-shell (either in the Makefile or as a `#!nix-shell` file), which is something I think we shouldn't do. _(sidebar: I don't think we should **force** users to run with nix though, nix is here for _our_ use and is our "known good dev env" but they should be able to use w/e they have as long as the binaries are around, ci will catch any version issues.)_ Then if you add that its a more pain to manage to shell.nix files than one it doesn't feel like its all that useful to split up.

> It might be a good idea to also move all the iPXE build related stuff in the current `Makefile` to the same location and then just import that into the top level `Makefile`. I feel like that could break down the complexity of the different moving parts of this repo. Thoughts?

I don't think I quite get what you're thinking here. If we move the rules to build/scripts/ipxe.mk and `include` the file not a whole lot changes. The files are split yeah and thats definitely nicer but the build is still the same. The other options is to have a totally separate Makefile in build/scripts and run something like `+$(MAKE) -C build/scripts build` or something. That would be `recursive-make` and this is "considered harmful".

**Comment by jacobweinstock** [comment]
> > Would you be open to leaving the `shell.nix` in its current location (`binary/script/shell.nix`)? It's only needed for iPXE builds.
> 
> An earlier version of this PR was like that, but it was still adding a top-level shell.nix file with `curl gnumake gnused go`. If we did the 2 shell.nix files we'd still need gnumake and go, and running the `fetch_and_extract_ipxe.sh` script should be done using nix-shell (either in the Makefile or as a `#!nix-shell` file), which is something I think we shouldn't do. _(sidebar: I don't think we should **force** users to run with nix though, nix is here for _our_ use and is our "known good dev env" but they should be able to use w/e they have as long as the binaries are around, ci will catch any version issues.)_ Then if you add that its a more pain to manage to shell.nix files than one it doesn't feel like its all that useful to split up.
> 
> > It might be a good idea to also move all the iPXE build related stuff in the current `Makefile` to the same location and then just import that into the top level `Makefile`. I feel like that could break down the complexity of the different moving parts of this repo. Thoughts?
> 
> I don't think I quite get what you're thinking here. If we move the rules to build/scripts/ipxe.mk and `include` the file not a whole lot changes. The files are split yeah and thats definitely nicer but the build is still the same. The other options is to have a totally separate Makefile in build/scripts and run something like `+$(MAKE) -C build/scripts build` or something. That would be `recursive-make` and this is "considered harmful".

I don't think we should have a top level shell.nix at all. It's not needed. curl and make are table stakes. 

My other comment about having a separate Makefile in the same ipxe location was only for abstracting complexity between the Go stuff and the iPXE building. It was only a thought. I'm fine with out it.

**Comment by mmlb** [comment]
> I don't think we should have a top level shell.nix at all. It's not needed. curl and make are table stakes.

Go would be in the top-level shell.nix file too.

I don't think we should be making assumptions about what the user has installed. Yes curl is probably around and is safe to rely, but make could be bsd make vs gnu make, they have some incompatibilities which I don't know of the top of my head (tbf I don't think we're likely to run into any of them either).

**Review by jacobweinstock — CHANGES_REQUESTED** [review]
I don't believe we should be adding more nix. I have heard too many ppl in the community bring up concerns with the heavy use of nix in the Tinkerbell project in general. I think the changes here outside of the top level nix file are fine though.

**Comment by mmlb** [comment]
> I don't believe we should be adding more nix. I have heard too many ppl in the community bring up concerns with the heavy use of nix in the Tinkerbell project in general. I think the changes here outside of the top level nix file are fine though.

Hoisting the shell.nix stuff up to the top-level doesn't make things any worse for non-nix users. The ipxe build script still requires nix to build like before (which we should fix too imo, as long as CI is the only place where ipxe is built). If someone doesn't have nix installed then shell.nix in the top-level means nothing to them.

I suspect some of the pain of nix is due to lack of alternative for non-nixers. Other tinkerbell projects make the mistake of having nix-shell in script shebangs, forcing the use of nix. Instead we should run the script using nix-shell or just use nix-shell to setup the PATH and run the script like normal. This would leave the chance that a users' local binaries are different versions and will still fail compared to CI but we should also add some info in the docs about possibly using nix-shell via docker.

File: shell.nix [hunk 1]
```diff
@@ -0,0 +1,18 @@
+let _pkgs = import <nixpkgs> { };
```
Comment (jacobweinstock) [inline shell.nix:1]: let's move this back into the `binary/` directory, update `IPXE_NIX_SHELL:=shell.nix` -> `IPXE_NIX_SHELL:=binary/shell.nix` in the `Makefile` and get this PR merged.

File: .github/workflows/ipxe.yaml [hunk 2]
```diff
@@ -17,4 +17,4 @@ jobs:
       - name: Build iPXE
         env:
           GITHUB_TOKEN: ${{ secrets.Token }}
-        run: (cd binary/; ./script/build_and_pr.sh)
+        run: make binary
```
[the remark this answers is not in the document]
Comment (mmlb) [reply .github/workflows/ipxe.yaml]: fixed this

File: shell.nix (hunk 1 above)
Comment (mmlb) [reply shell.nix]: I've put the shell.nix file back, still don't like it though.

**Comment by mmlb** [comment]
Hey @jacobweinstock I've re-updated the nixpkgs revision and kept the file where it was. Anything else you'd like to see here? Looking to get this merged soon hopefully.

**Comment by jacobweinstock** [comment]
Hey @mmlb, thanks for the update. I'm getting an error with `make binary`

```bash
$ make binary
"9062544f6a0c69c249b90d21a08d05518aafc2ec"
sh: line 1: 9062544f6a0c69c249b90d21a08d05518aafc2ec: command not found
make: *** [Makefile:25: upstream-9062544f6a0c69c249b90d21a08d05518aafc2ec/README] Error 127
```
Looks like the variable `IPXE_FETCH_SCRIPT` from line 25 in the Makefile isn't set anywhere?

**Comment by mmlb** [comment]
@jacobweinstock whoops rebase fail. I've got it fixed up but now am thinking we should build the binaries as part of ci always (but not commit) to ensure that code doesn't bit rot. Does that sound good to you?

**Comment by jacobweinstock** [comment]
hmmm.....building iPXE every time increases build times significantly. I'm leaning toward not building every time. Currently, iPXE will build for PR's that change direct the iPXE binary in any way. This should already catch build issues. I can understand the desire to catch any build tooling updates. I just don't know if i'm fully convinced on the increased build time trade off.

**Comment by mmlb** [comment]
sigh... yeah. Foiled again by GHA's low core count (on my machine I fetch and build in total of 22.57s). Dropping that then.

**Comment by mmlb** [comment]
@jacobweinstock PTAL again

File: binary/script/shell.nix [hunk 3]
```diff
@@ -1,27 +1,31 @@
...
-
-in mkShell {
-  buildInputs = [
-    curl
-    expect
-    gcc
-    git
-    gnumake
-    go
-    perl
```
Comment (jacobweinstock) [inline binary/script/shell.nix:22]: Looks like `perl` is still needed. 
```bash
sh: line 1: perl: command not found
make: *** [Makefile.housekeeping:1221: bin-x86_64-efi/ipxe.efi.tmp] Error 127
rm bin-x86_64-efi/version.ipxe.efi.o
make: Leaving directory '/code/upstream-9062544f6a0c69c249b90d21a08d05518aafc2ec-ipxe.efi/src'
make: *** [Makefile:30: binary/ipxe.efi] Error 2
```

Comment (mmlb) [reply binary/script/shell.nix:22]: oops I thought that was for `git` :man_facepalming:

File: Makefile [hunk 4]
```diff
@@ -1,30 +1,39 @@
...
+IPXE_BUILD_SCRIPT := binary/script/build_ipxe.sh
+IPXE_FETCH_SCRIPT := binary/script/fetch_and_extract_ipxe.sh
+IPXE_NIX_SHELL := binary/script/shell.nix
 
 help: ## show this help message
 	@grep -E '^[a-zA-Z_-]+.*:.*?## .*$$' Makefile | sort | awk 'BEGIN {FS = ":.*?## "}; {printf "\033[32m%-30s\033[0m %s\n", $$1, $$2}'
 
 include lint.mk
 
+.PHONY: binary
```
Comment (jacobweinstock) [inline Makefile:12]: sorry, just caught this now. Is this wanted? `make binary` is running every time for me.

**Review by jacobweinstock — COMMENTED** [review]
With `make binary` when ,what I believe is, a successful build completes it appears that the return code is non zero.

```bash
rm bin-arm64-efi/version.snp.efi.o
make: Leaving directory '/code/upstream-9062544f6a0c69c249b90d21a08d05518aafc2ec-snp.efi/src'
+ cp -a upstream-9062544f6a0c69c249b90d21a08d05518aafc2ec-snp.efi/src/bin-arm64-efi/snp.efi binary/snp.efi

[nix-shell:/tmp/ipxedust]$ echo $?
2
```

File: Makefile (hunk 4 above)
Comment (mmlb) [reply Makefile:12]: we do want this because it is a phony target, it should be ok because nothing calls `make binary` unless we know we need to rebuild. Its rebuilding for you because `make binary` isn't succeeding. I'm not sure why its not working for you can you post the full make output? 

"Works on my machine"
```
[16:49:55]-[~/g/s/g/t/ipxedust]-[manny@zennix]
git clean -fxd; make binary &>/dev/null; echo status=$status
status=0
[16:50:32]-[~/g/s/g/t/ipxedust]-[manny@zennix]
make binary
make: Nothing to be done for 'binary'.
```

Comment (jacobweinstock) [reply Makefile:12]: ahh...ok. that make sense, thanks. Let me do another run and see.

Comment (jacobweinstock) [reply Makefile:12]: Maybe i had a different commit because it's working now. Thanks!

**Review by jacobweinstock — APPROVED** [review]
Thanks for this!
